File: python/meetup/meetup.py
from datetime import date
import calendar

DAY_NAMES_INDICES = dict([(calendar.day_name[x], x) for x in range(7)])

def meetup_day(year, month, day_of_the_week, which):
    cal = calendar.monthcalendar(year, month)
    dates_ = []

    for week in cal:
        day_index = DAY_NAMES_INDICES[day_of_the_week]
        date_ = week[day_index]
        if date_ > 0:
            dates_.append(date_)

    if which == 'last':
        return date(year, month, max(dates_))

    for index, val in enumerate(dates_):
        if which == 'teenth' and val > 12:
            return date(year, month, val)
        elif which != 'teenth' and index == int(which[0]) - 1:
            return date(year, month, val)
    raise Exception()

File: elixir/lib/meetup/meetup.ex
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
          :monday
          | :tuesday
          | :wednesday
          | :thursday
          | :friday
          | :saturday
          | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @weekday_names_map [:monday, :tuesday, :wednesday, :thursday, :friday, :saturday, :sunday]
                     |> Enum.with_index(1)
                     |> Enum.into(%{})

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date()
  def meetup(year, month, weekday, schedule) do
    {:ok, first_date_in_month} = Date.new(year, month, 1)
    {day_of_month, weekday_date} = weekday_date(first_date_in_month, weekday)

    case schedule do
      :first ->
        weekday_date

      :second ->
        Date.add(weekday_date, 7)

      :third ->
        Date.add(weekday_date, 14)

      :fourth ->
        Date.add(weekday_date, 21)

      :last ->
        last_day = day_of_month + 28

        last_day =
          if last_day > Date.days_in_month(weekday_date) do
            last_day - 7
          else
            last_day
          end

        Date.add(weekday_date, last_day - day_of_month)

      :teenth ->
        Date.add(weekday_date, teenth(day_of_month + 7) - day_of_month)
    end
    |> Date.to_erl()
  end

  defp teenth(day) when day > 12, do: day
  defp teenth(day), do: teenth(day + 7)

  defp weekday_date(%Date{} = start_of_month_date, weekday_name) do
    weekday = @weekday_names_map[weekday_name]

    case Date.day_of_week(start_of_month_date) do
      ^weekday ->
        {1, start_of_month_date}

      first_weekday when first_weekday > weekday ->
        days = 7 - first_weekday + weekday
        {days + 1, Date.add(start_of_month_date, days)}

      first_weekday ->
        days = weekday - first_weekday
        {days + 1, Date.add(start_of_month_date, days)}
    end
  end
end
